src/main/java/de/philippkatz/maven/plugins/TestParserMojo.java

```diff
@@ -86,6 +86,41 @@ public void execute() throws MojoExecutionException {
 			throw new MojoExecutionException("There were test failures.");
 		}
 	}
+	
+	/**
+	 * Walks over a tree of directories and recursively adds all these directories to a list of <code>File</code> objects.
+	 * 
+	 * @param rootDirectory The root directory from where to start the search for sub directories.
+	 * @return A <code>List</code> of the directory provided as root and all its sub directories.
+	 * @throws MojoExecutionException 
+	 */
+	List<File> collectReportDirectoriesRecursively(final File rootDirectory) throws MojoExecutionException {
+		if(rootDirectory == null) {
+			throw new MojoExecutionException("No valid directory provided");
+		}
+		if(!rootDirectory.exists()) {
+			throw new MojoExecutionException("Directory "+rootDirectory+" does not exist!");
+		}
+		if(!rootDirectory.isDirectory()) {
+			throw new MojoExecutionException("Directory "+rootDirectory+" is no directory!");
+		}
+		
+		List<File> ret = new ArrayList<>();
+		ret.add(rootDirectory);
+		File[] subDirectories = rootDirectory.listFiles(new FileFilter() {
+			
+			@Override
+			public boolean accept(final File pathname) {
+				return pathname.isDirectory();
+			}
+		});
+		
+		for (File dir: subDirectories) {
+			ret.addAll(collectReportDirectoriesRecursively(dir));
+		}
```

Comment:
Having above's lines starting at 110 in a single loop would be much easier to grasp:

    for (File file : rootDirectory.listFiles()) {
      if (file.isDirectory()) {
        ret.addAll(collectReportDirectoriesRecursively(file));
      }
    }